### DISCUSSION TITLE
Formatted .JSON files from PBI to Files

### DISCUSSION THREAD

**Discussion by hlynurarni** [didierterrien/Power-BI-Sidetools#27] [Ideas]
Would it be possible to have the .json files correctly formatted ?

I am trying to avoid using git hooks before commit. The files cant be diffed in a meaningful way if they are not formatted.

This is what I get: 

![image](https://user-images.githubusercontent.com/10733420/129006866-3b9f55d8-62ed-4173-a22b-1697a41751a5.png)

This is what I want:

![image](https://user-images.githubusercontent.com/10733420/129006984-0b79643a-cf07-4379-adad-5c44bf4cfcbe.png)

**Comment by didierterrien** [comment]
It seems it's not hard to do. I will investigate further soon.
By the way, did you start source control in production ?

**Reply by hlynurarni** [reply]
We have not rolled it out in production yet, will happen next week hopefully (vacation days are over). Thank you for the quick responses and being so open minded about additions.

Not sure if this is already provided but could be available through SideTools ? 

- Incremental refresh configuration
- Row level security configuration

**Reply by didierterrien** [reply]
The whole model is exposed to Sidetools so everything can be potentially available. These 2 features were not on the roadmap but I was thinking about exposing the whole BIM file. I need to investigate further.

**Reply by didierterrien** [reply]
Could you please explain what you expect in these requests ?
Do you need only the DAX formula in RLS filter expression ?
About incremental refresh, please send Tabular Editor screenshots.

**Reply by didierterrien** [reply]
I added RLS filter expressions in version 1.2.5
I don't understand what you mean by "Incremental refresh configuration" but I added also the Model.bim file in version 1.2.5
The Model.bim file is a Json file containing the whole model definition.

**Reply by hlynurarni** [reply]
Hey sorry for the silence, 

Model.bim is really nice to have, great effort! Will look into what I can infer or gain from the file that we don't already have.

Incremental refresh was mainly to track if there was any change made in regards to that. 
RLS filter expression, thank you ! (hopefully Microsoft will provide and API to add new emails to roles)

Will try out the updated 1.2.5 now and give feedback if needed.

When this is in production I will give you an update of the overall process. Think this tool should get way more recognition then it does and especially how quick and open you are for improvements and discussion. Thanks again!

**Reply by didierterrien** [reply]
Hello,
Version 1.2.5 is not published yet. I will publish it this week. Please do not try the test version.
Your feedback is very important for the community as source control is one of the most requested feature by professionals. There are several attempts on social media but none seems very robust.
Thanks again for your contribution and propositions. It's a pleasure to collaborate.

**Reply by hlynurarni** [reply]
I get it and likewise on the collaboration.

FYI I found the tool through your comment on the section of this video: [Source control, Guy in a cube](https://www.youtube.com/watch?v=fCGd5qID-dY&ab_channel=GuyinaCube), find it baffling that they didn't mention this tool in the video. 

![image](https://user-images.githubusercontent.com/10733420/130753037-9d8a4439-61ea-4def-99af-992f11ac2c02.png)

Since it had so little traction, as it seems, I was skeptical to begin with. To me, SideTools is the clear option to go with (spent some decent time on looking through all options)

**Reply by didierterrien** [reply]
It seems "Sidetools" is a forbidden word in many mouths ;)  Many people are aware about it (Marco Russo, Daniel Otykier, Guy in the Cube, Mike Carlo, ...) but still, it's not mentioned. That's a mystery for me. Anyway, I have good contacts with some MVPs so I progress slowly. 
If you manage to implement a very good source control solution, you may publish a post and I will promote it. I may even ask to an MVP to showcase it if you want. 
I have huge hopes about the next features I will publish. One of them is DAX trainer (you can check the mockup in GitHub discussions). The second one is still secret.   
Thanks a lot.

**Comment by didierterrien** [comment]
Done. It will appear in version 1.2.5
I added a setting to optionally format the Json. By default, it is set to false in order to keep the previous behaviour.
I will close the issue when version 1.2.5 will be published.
Thank you for your contribution.

**Reply by hlynurarni** [reply]
Wow that was quick, thank you so much ! Great job.

**Comment by didierterrien** [comment]
Power BI Sidetools has been updated to version 1.2.5

**Comment by hlynurarni** [comment]
I am trying out the new version and added features for version 1.2.5

The json formatting option works great but it seems that I am no longer able to get the Layouts for my live reports as I could before. I do get the folder when I do "pbi to files" for my data_model.pbix.

![image](https://user-images.githubusercontent.com/10733420/132000138-d2f10a21-30a8-494d-ad58-4596c3a3abdd.png)

The files in red bracket are no longer available.

The log file is here: 
[Power BI Sidetools log.txt](https://github.com/didierterrien/Power-BI-Sidetools/files/7105649/Power.BI.Sidetools.log.txt)

**Reply by didierterrien** [reply]
I feel this issue is linked to #26 
Do you confirm DAX and M files are correctly  created ? I suppose not. The problem seems to be connected to the live connection which is not supported yet.

**Reply by hlynurarni** [reply]
There are no DAX or M files in the .pbix, but the folders are created empty, which is fine. All our reports are live (excluding data models) and we got the Layout files to track changes made to those reports. Now that option seems to be gone with the update 😕

**Reply by didierterrien** [reply]
I will have a look to solve your issue but I'm not sure to have enough time to test is properly in a small amount of time. I will publish a test version so you can try it

**Reply by hlynurarni** [reply]
Thanks for all the effort, I just report as soon as I see some unexpected changes or bugs.

**Comment by didierterrien** [comment]
I hope I solved the issue in V1.2.6. This test version will be available soon

**Reply by hlynurarni** [reply]
For reference the layout folder and files are created with a test version of v1.2.5 for live reports.

[Power-BI-Sidetools V1.2.5 test.zip](https://github.com/didierterrien/Power-BI-Sidetools/files/7169497/Power-BI-Sidetools.V1.2.5.test.zip)  (Downloaded at 7.19.2021)

This thread is getting too long to follow (discussions are in multiple unrelated places at the moment). Consider to mark this page as "closed", "solved", "enriched". 

This .json format feature is added and works great !

Just wanted to give you the update.
